metricbeat/module/elasticsearch/index/data.go

```diff
@@ -19,70 +19,379 @@ package index
+	Evictions         int `json:"evictions"`
+	HitCount          int `json:"hit_count"`
+	MissCount         int `json:"miss_count"`
+}
+
+type shardStats struct {
+	Total     int `json:"total"`
+	Primaries int `json:"primaries"`
+	Replicas  int `json:"replicas"`
+
+	ActiveTotal     int `json:"active_total"`
+	ActivePrimaries int `json:"active_primaries"`
+	ActiveReplicas  int `json:"active_replicas"`
+
+	UnassignedTotal     int `json:"unassigned_total"`
+	UnassignedPrimaries int `json:"unassigned_primaries"`
+	UnassignedReplicas  int `json:"unassigned_replicas"`
+
+	Initializing int `json:"initializing"`
+	Relocating   int `json:"relocating"`
+}
+
+type bulkStats struct {
+	TotalOperations   int `json:"total_operations"`
+	TotalTimeInMillis int `json:"total_time_in_millis"`
+	TotalSizeInBytes  int `json:"total_size_in_bytes"`
+	AvgTimeInMillis   int `json:"avg_time_in_millis"`
+	AvgSizeInBytes    int `json:"avg_size_in_bytes"`
+}
+
+func eventsMapping(r mb.ReporterV2, m *MetricSet, info elasticsearch.Info, content []byte) error {
+	clusterStateMetrics := []string{"metadata", "routing_table"}
+	clusterState, err := elasticsearch.GetClusterState(m.HTTP, m.HTTP.GetURI(), clusterStateMetrics)
+	if err != nil {
+		return errors.Wrap(err, "failure retrieving cluster state from Elasticsearch")
 	}
-)
 
-func eventsMapping(r mb.ReporterV2, info elasticsearch.Info, content []byte) error {
-	var indicesStruct IndicesStruct
-	err := json.Unmarshal(content, &indicesStruct)
+	var indicesStats stats
+	if err := parseAPIResponse(content, &indicesStats); err != nil {
+		return errors.Wrap(err, "failure parsing Indices Stats Elasticsearch API response")
+	}
+
+	indicesSettings, err := elasticsearch.GetIndicesSettings(m.HTTP, m.HTTP.GetURI())
 	if err != nil {
-		return errors.Wrap(err, "failure parsing Elasticsearch Stats API response")
+		return errors.Wrap(err, "failure retrieving indices settings from Elasticsearch")
 	}
 
 	var errs multierror.Errors
-	for name, index := range indicesStruct.Indices {
-		event := mb.Event{}
+	for name, idx := range indicesStats.Indices {
+		event := mb.Event{
+			ModuleFields: common.MapStr{},
+		}
+		idx.Index = name
 
-		event.RootFields = common.MapStr{}
-		event.RootFields.Put("service.name", elasticsearch.ModuleName)
+		settings, exists := indicesSettings[name]
+		if exists {
+			idx.Hidden = settings.Hidden
+		}
+
+		err = addClusterStateFields(&idx, clusterState)
+		if err != nil {
+			errs = append(errs, errors.Wrap(err, "failure adding cluster state fields"))
+			continue
+		}
 
-		event.ModuleFields = common.MapStr{}
-		event.ModuleFields.Put("cluster.name", info.ClusterName)
 		event.ModuleFields.Put("cluster.id", info.ClusterID)
+		event.ModuleFields.Put("cluster.name", info.ClusterName)
 
-		event.MetricSetFields, err = schema.Apply(index)
+		// Convert struct to common.Mapstr by passing it to JSON first so we can store the data in the root of the
+		// metricset level
+		byt2, err := json.Marshal(idx)
```

Comment:
Consider replacing `byt2` with something more descriptive or at least less arbitrary?

Reply:
Yeah, good catch, I change its name to something more meaningful